### ISSUE TITLE
Heroic Detection still not working on flathub on Steam Deck

### ISSUE THREAD

**Issue by jimbojd72** [mtkennerly/ludusavi#186]
### Ludusavi version

v0.15.2

### Operating system

Linux (Steam Deck)

### Installation method

Flatpak

### Description

I know it is supposed to worked based on https://github.com/mtkennerly/ludusavi/pull/141

Configuring /home/deck/.var/app/com.heroicgameslauncher.hgl/config/heroic root manually doesn't find anything

I put pictures of my configurations and an samples of the games I have installed just in case.

### Logs

![image](https://user-images.githubusercontent.com/29709551/222775727-9fbf4570-d7c1-4902-80e2-36a7511f1fa3.png)
![image](https://user-images.githubusercontent.com/29709551/222775998-0de95fbf-32d3-48b3-9f48-114e283ad07d.png)

**Comment by jimbojd72** [comment]
[ludusavi_rCURRENT.log](https://github.com/mtkennerly/ludusavi/files/10884410/ludusavi_rCURRENT.log)

config.yaml
```
---
manifest:
  url: "https://raw.githubusercontent.com/mtkennerly/ludusavi-manifest/master/data/manifest.yaml"
  etag: ~
language: en-US
theme: dark
roots:
  - path: /home/deck/.local/share/Steam
    store: steam
  - path: /run/media/mmcblk0p1
    store: steam
  - path: /home/deck/.var/app/com.heroicgameslauncher.hgl/config/heroic
    store: heroic
redirects: []
backup:
  path: /home/deck/ludusavi-backup
  ignoredGames: []
  merge: true
  filter:
    ignoredPaths: []
    ignoredRegistry: []
  toggledPaths: {}
  toggledRegistry: {}
  sort:
    key: name
    reversed: false
  retention:
    full: 1
    differential: 2
  format:
    chosen: zip
    zip:
      compression: zstd
    compression:
      deflate:
        level: 6
      bzip2:
        level: 6
      zstd:
        level: 10
restore:
  path: /home/deck/ludusavi-backup
  ignoredGames: []
  sort:
    key: name
    reversed: false
customGames: []

```

**Comment by mtkennerly** [comment]
Thanks for the detailed report!

Of those four games, PCGW doesn't have any info for Demon Turf save data, but the other three should work. I see that FTL and Loop Hero have native Linux versions (I'm assuming Heroic would install these instead of the Windows versions), and FIST only has a Windows version available.

Could you double check a few things?

* For FTL, does this folder exist? `/home/deck/.var/app/com.heroicgameslauncher.hgl/data/FasterThanLight`
* For FIST, there should be a corresponding JSON file In `/home/deck/.var/app/com.heroicgameslauncher.hgl/config/heroic/GamesConfig`. What's the `winePrefix` setting in that file?
* Inside of FIST's `winePrefix` folder, there should be these subfolders: `users/*/AppData/Local/ZingangGame/Saved/SaveGames` (where `*` is any username). Does that exist?

**Comment by jimbojd72** [comment]
I don't think this is a game specific issue but a Heroic parser issue. But with your question I was able to dig deeper

I look into `/home/deck/.var/app/com.heroicgameslauncher.hgl/data` like suggested but there is almost nothing:

```
(130)(deck@steamdeck ~)$ ls /home/deck/.var/app/com.heroicgameslauncher.hgl/data/FasterThanLight
ls: cannot access '/home/deck/.var/app/com.heroicgameslauncher.hgl/data/FasterThanLight': No such file or directory
(2)(deck@steamdeck ~)$ ls /home/deck/.var/app/com.heroicgameslauncher.hgl/data/
applications  Steam
(deck@steamdeck ~)$ ls /home/deck/.var/app/com.heroicgameslauncher.hgl/data/applications/
(2)(deck@steamdeck ~)$ ls /home/deck/.var/app/com.heroicgameslauncher.hgl/data/Steam
```

I noticed my games are in fact in `/home/deck/Games/Heroic` and prefixed in `/home/deck/Games/Heroic/Prefixes`

```
(deck@steamdeck ~)$ ls /home/deck/Games/Heroic
 AxiomVerge2    EnterTheGungeon   Flashback                 HeroicBashLauncher-2.9.2   LoopHero        Prefixes      TheEndIsNigh
 DemonTurf      FasterThanLight  'Greak Memories of Azur'   InscryptionWkwPQ           NuclearThrone   RogueLegacy   VigilTheLongestNight
 DiscoElysium   FIST              Hades                     IntoTheBreach              Postal         'Samorost 1'
(deck@steamdeck ~)$ ls /home/deck/Games/Heroic/Prefixes/
 AxiomVerge2   DiscoElysium         FasterThanLight                    Hades             'Into The Breach'   LoopHero
'Demon Turf'   EnterTheGungeon     'F.I.S.T. Forged In Shadow Torch'   Inscryption        IntoTheBreach      NuclearThrone
 DemonTurf    'Faster Than Light'  'Greak Memories of Azur'            InscryptionWkwPQ  'Loop Hero'         TheEndIsNigh
```

But path `/home/deck/Games/Heroic`  should be discoverable from heroic config file for ludusavi to find as you can see below:

```
(deck@steamdeck heroic)$ cat /home/deck/.var/app/com.heroicgameslauncher.hgl/config/heroic/config.json
{
  "defaultSettings": {
    "checkUpdatesInterval": 10,
    "enableUpdates": false,
    "addDesktopShortcuts": false,
    "addStartMenuShortcuts": false,
    "autoInstallDxvk": false,
    "autoInstallVkd3d": false,
    "addSteamShortcuts": true,
    "preferSystemLibs": false,
    "checkForUpdatesOnStartup": false,
    "autoUpdateGames": false,
    "customWinePaths": [],
    "defaultInstallPath": "/home/deck/Games/Heroic",   <<<<---------- Interesting path to parse
    "libraryTopSection": "recently_played_installed",
    "defaultSteamPath": "/home/deck/.steam/steam",
    "defaultWinePrefix": "/home/deck/Games/Heroic/Prefixes", <<<<---------- Interesting path to parse
    "hideChangelogsOnStartup": false,
    "language": "en",
    "maxWorkers": 0,
    "minimizeOnLaunch": false,
    "nvidiaPrime": false,
    "enviromentOptions": [],
    "wrapperOptions": [],
    "showFps": false,
    "useGameMode": false,
    "userInfo": {
      "epicId": "a5a09135dee444b3a25747cc694d2e3d",
      "name": "deck"
    },
    "wineCrossoverBottle": "Heroic",
    "winePrefix": "/home/deck/.wine",
    "wineVersion": {
      "bin": "/home/deck/.local/share/Steam/steamapps/common/Proton 7.0/proton",
      "name": "Proton - Proton 7.0",
      "type": "proton"
    },
    "altLegendaryBin": "",
    "altGogdlBin": "",
    "disableController": false,
    "egsLinkedPath": "",
    "maxRecentGames": 5,
    "showUnrealMarket": false,
    "eacRuntime": false,
    "battlEyeRuntime": false
  },
  "version": "v0"
}
```

Long story short: I guess ludusavi should get the default paths from `/home/deck/.var/app/com.heroicgameslauncher.hgl/config/heroic/config.json` and add those to the roots to backup.

**Comment by mtkennerly** [comment]
There should be some game-specific config files that identify which prefix belongs to each game. That's what Ludusavi looks for right now, so I'd like to figure out why that part isn't working. What do you see in `/home/deck/.var/app/com.heroicgameslauncher.hgl/config/heroic/GamesConfig` ? Could you share some of the files from that folder, particularly for FTL and FIST?

**Comment by jimbojd72** [comment]
No idea what is the naming scheme of those files, but with some grep I found:

```
(deck@steamdeck GamesConfig)$ cat /home/deck/.var/app/com.heroicgameslauncher.hgl/config/heroic/GamesConfig/bf4808b3395b43fb80afbe6773e1661f.json
{
  "bf4808b3395b43fb80afbe6773e1661f": {
    "autoInstallDxvk": false,
    "autoInstallVkd3d": false,
    "preferSystemLibs": false,
    "nvidiaPrime": false,
    "enviromentOptions": [],
    "wrapperOptions": [],
    "showFps": false,
    "useGameMode": false,
    "language": "",
    "wineVersion": {
      "bin": "/home/deck/.local/share/Steam/steamapps/common/Proton 7.0/proton",
      "name": "Proton - Proton 7.0",
      "type": "proton"
    },
    "winePrefix": "/home/deck/Games/Heroic/Prefixes/F.I.S.T. Forged In Shadow Torch"
  },
  "version": "v0",
  "explicit": true
}
```
FTL:

```
(deck@steamdeck GamesConfig)$ cat /home/deck/.var/app/com.heroicgameslauncher.hgl/config/heroic/GamesConfig/Blowfish.json 
{
  "Blowfish": {
    "autoInstallDxvk": false,
    "autoInstallVkd3d": false,
    "preferSystemLibs": false,
    "maxSharpness": 2,
    "nvidiaPrime": false,
    "enviromentOptions": [],
    "wrapperOptions": [],
    "savesPath": "",
    "showFps": false,
    "targetExe": "",
    "useGameMode": false,
    "language": "",
    "wineVersion": {
      "bin": "/home/deck/.local/share/Steam/steamapps/common/Proton 7.0/proton",
      "name": "Proton - Proton 7.0",
      "type": "proton"
    },
    "winePrefix": "/home/deck/Games/Heroic/Prefixes/Faster Than Light",
    "eacRuntime": false,
    "battlEyeRuntime": false,
    "gogSaves": [],
    "wineCrossoverBottle": ""
  },
  "version": "v0",
  "explicit": true
}
```

**Comment by mtkennerly** [comment]
Yeah, the naming is usually weird 😅 Thanks, the contents there look normal.

Would you mind giving me the log for this command for some extra details?

```
flatpak run --env=RUST_LOG=ludusavi=trace com.github.mtkennerly.ludusavi backup --preview "
FTL: Faster Than Light"
```

Here's what Ludusavi does for Epic games (I'll use `${heroic}` as an abbreviation for `/home/deck/.var/app/com.heroicgameslauncher.hgl`):

* Check `${heroic}/config/legendary` and/or `~/.config/legendary` for a file named `installed.json`.
  * From that file, record each game's `app_name` and `title`.
* Check `${heroic}/config/heroic/GamesConfig` for a JSON file with the same `app_name` from the above step.
  * From that file, record the game's `wine_prefix`.
* Finally, the `title` needs to match the PCGamingWiki article name, and then Ludusavi can connect the game to its prefix.

If that works, then the next question is whether the saves exist where Ludusavi expects. For FTL, based on your `Blowfish.json`, that would be:

* `/home/deck/Games/Heroic/Prefixes/Faster Than Light/users/*/Documents/My Games/FasterThanLight`

Does that exist?

**Comment by jimbojd72** [comment]
> Would you mind giving me the log for this command for some extra details?

Not a lot here:

```
(2)(deck@steamdeck ~)$ flatpak run --env=RUST_LOG=ludusavi=trace com.github.mtkennerly.ludusavi backup --preview "FTL: Faster Than Light"

Overall:
  Games: 0
  Size: 0 B
  Location: /home/deck/ludusavi-backup
```

> /home/deck/Games/Heroic/Prefixes/Faster Than Light/drive_*/users/*/Documents/My Games/FasterThanLight
```
(2)(deck@steamdeck ~)$ ls /home/deck/Games/Heroic/Prefixes/Faster\ Than\ Light/
config_info  pfx  pfx.lock  tracked_files  version
```

There is a missing `pfx` directory in your path above. But I could find the saves under this path instead:
```
(deck@steamdeck ~)$ ls /home/deck/Games/Heroic/Prefixes/Faster\ Than\ Light/pfx/drive_c/users/steamuser/Documents/My\ Games/FasterThanLight/
ae_prof.sav  continue.sav
```

**Comment by jimbojd72** [comment]
I think you are already fine but added the requested log file just in case.

[ludusavi_rCURRENT.log](https://github.com/mtkennerly/ludusavi/files/10893037/ludusavi_rCURRENT.log)


Thanks for being responsive!

**Comment by djibux** [comment]
@jimbojd72, did version 0.16 fix things for you? I am still experiencing the issue.

**Comment by mtkennerly** [comment]
@djibux Could you give me some info so I can double check what's happening?

* A screenshot of the roots you have configured
* An example of a game that's not working
* Run this command (replace `FTL: Faster Than Light` with the game above):

  ```
  flatpak run --env=RUST_LOG=ludusavi=trace com.github.mtkennerly.ludusavi backup --preview "FTL: Faster Than Light"
  ```

  Then post your `~/.var/app/com.github.mtkennerly.ludusavi/config/ludusavi/ludusavi_rCURRENT.log` file.

**Comment by mtkennerly** [comment]
Strangely, it's not even checking the Heroic root. The only way that should happen is if there were an error when evaluating the root path as a glob (maybe from permissions). I can upload a beta Flatpak build to add some extra logging to confirm, but in the meantime, what permissions do you see from `ls -l /home/deck/.var/app/com.heroicgameslauncher.hgl/config/heroic` ?

**Comment by djibux** [comment]
Hello,
Here is the result of the ls command.
I didn’t mess with permissions, basically it’s a clean install with just Heroic and ludusavi.

```bash
total 288
-rw-r--r-- 1 deck deck 115627 Mar 19 21:42  areweanticheatyet.json
drwx------ 3 deck deck   4096 Mar 19 21:42  blob_storage
drwx------ 3 deck deck   4096 Mar  1 10:51  Cache
drwx------ 4 deck deck   4096 Mar  1 10:51 'Code Cache'
-rw-r--r-- 1 deck deck   1244 Mar  2 22:16  config.json
-rw------- 1 deck deck  20480 Mar  1 20:36  Cookies
-rw------- 1 deck deck      0 Mar  1 20:36  Cookies-journal
drwx------ 2 deck deck   4096 Mar  1 10:51  Crashpad
drwx------ 2 deck deck   4096 Mar  1 10:51  DawnCache
drwx------ 2 deck deck   4096 Mar  1 10:51  Dictionaries
drwxr-xr-x 2 deck deck   4096 Mar 18 02:35  GamesConfig
drwxr-xr-x 2 deck deck   4096 Mar 19 21:42  gog_store
drwx------ 2 deck deck   4096 Mar  1 10:51  GPUCache
drwxr-xr-x 2 deck deck   4096 Mar 18 02:34  icons
drwxr-xr-x 2 deck deck  57344 Mar 18 02:36  images-cache
drwxr-xr-x 2 deck deck   4096 Mar 19 21:42  lib-cache
drwx------ 3 deck deck   4096 Mar  1 10:51 'Local Storage'
drwx------ 2 deck deck  12288 Mar 19 21:42  logs
-rw------- 1 deck deck   1830 Mar 18 10:30 'Network Persistent State'
drwx------ 3 deck deck   4096 Mar  1 10:54  Partitions
-rw------- 1 deck deck     57 Mar  1 10:51  Preferences
drwx------ 2 deck deck   4096 Mar 19 21:54 'Session Storage'
drwxr-xr-x 2 deck deck   4096 Mar  2 22:16  sideload_apps
drwxr-xr-x 2 deck deck   4096 Mar 19 21:54  store
drwxr-xr-x 6 deck deck   4096 Mar  1 10:54  tools
-rw------- 1 deck deck    872 Mar  1 20:36  TransportSecurity
```

**Comment by mtkennerly** [comment]
Yeah, that does look normal. Thanks for checking.

I've published a beta build that's mostly the same as v0.16.0, but with some extra logging where I think the failure is happening. I'm not sure if the Discover app shows beta builds, but you can install  it from the terminal like this:

```
flatpak remote-add --if-not-exists flathub-beta https://flathub.org/beta-repo/flathub-beta.flatpakrepo
flatpak install flathub-beta com.github.mtkennerly.ludusavi
```

Then please run this command (note the `//beta`) and upload the new `ludusavi_rCURRENT.log`:

```
RUST_LOG=ludusavi=trace flatpak run com.github.mtkennerly.ludusavi//beta backup --preview "The Witness"
```

If needed, you can also switch your default version of Ludusavi this way:

```
flatpak make-current com.github.mtkennerly.ludusavi stable
flatpak make-current com.github.mtkennerly.ludusavi beta
```

**Comment by djibux** [comment]
Thanks for the detailed instructions.

Here is the attached log.

[ludusavi_rCURRENT.log](https://github.com/mtkennerly/ludusavi/files/11016230/ludusavi_rCURRENT.log)

**Comment by mtkennerly** [comment]
That's bizarre. It's not hitting any errors, but it's just acting like `/home/deck/.var/app/com.heroicgameslauncher.hgl/config/heroic/` doesn't exist.

Could you try a couple of small tests?

* If you add a new Heroic root and use the folder picker (<img width="57" alt="image" src="https://user-images.githubusercontent.com/4934192/226430273-f203dcfd-6cb5-4036-91e7-5e6b801e6b86.png">) to browse to `/home/deck/.var/app/com.heroicgameslauncher.hgl/config/heroic/` , does it still show the same path in Ludusavi? (You'll need to do this in desktop mode to use the folder picker.)
* This shouldn't make any difference, but what if you remove the slash at the end of the path?

I've also updated the beta with some additional logging as a sanity check. You can update with `flatpak update com.github.mtkennerly.ludusavi//beta`, and then please try The Witness one more time.

**Comment by djibux** [comment]
Adding using the folder has the same behavior.  Folder does not get renamed (and is added without the trailing slash).  I noticed folders on `/run/media/mmcblk0p1/` do get renamed though, so I understand you wanted to double check.

New beta version confirms that ludusavi does not see the folder.  I added it using flatseal and managed to get data for "The Witness" from the interface but not the other three games.

[ludusavi_rCURRENT.log (before using flatseal)](https://github.com/mtkennerly/ludusavi/files/11022959/ludusavi_rCURRENT.log)
`TRACE [ludusavi::config] Configured root (Heroic): /home/deck/.var/app/com.heroicgameslauncher.hgl/config/heroic | interpreted: /home/deck/.var/app/com.heroicgameslauncher.hgl/config/heroic | exists: false | is dir: false`

[ludusavi_rCURRENT_fs.log (after using flatseal)](https://github.com/mtkennerly/ludusavi/files/11022960/ludusavi_rCURRENT_fs.log)

Note: I removed Steam roots just to make sure they didn’t get in the way)

**Comment by djibux** [comment]
My bad sorry, it finds The Witcher after adding the directory in flatseal but not The Witness.  The Witcher is the only gog game, others are Epic.

**Comment by djibux** [comment]
Last comment and I shut up: if I add `/home/deck/.var/app/com.heroicgameslauncher.hgl` in flatseal, both The Witcher and The Witness are detected, but not Batman and Into The Breach.  I noticed in [your comment](https://github.com/mtkennerly/ludusavi/issues/186#issuecomment-1455189467) above that you also need access to `${heroic}/config/legendary`.

**Comment by mtkennerly** [comment]
That's a great find! Ludusavi's [Flatpak setup](https://github.com/flathub/com.github.mtkennerly.ludusavi/blob/master/com.github.mtkennerly.ludusavi.yaml) specifies:

```
  - --filesystem=home
  - --filesystem=/run/media/
  - --filesystem=/media/
  - --filesystem=/var/home/
```

I would have thought that the `home` permission would be enough here, but I found this note:

https://docs.flatpak.org/en/latest/sandbox-permissions.html#filesystem-access

> The `home` permission also has exceptions as it does not grant access to the subdirectories for other applications in `~/.var/app/`.

I've updated the beta to include `--filesystem=~/.var/app/com.heroicgameslauncher.hgl` (based on https://github.com/flatpak/flatpak/issues/2735#issuecomment-469499858). Could you try removing the Flatseal changes and see if the updated beta can find The Witness and The Witcher?

If that works, then I'll add the permissions to the stable Flatpak release as well, and then we can investigate Batman and Into the Breach specifically.

**Comment by mtkennerly** [comment]
That's a great find! Ludusavi's [Flatpak setup](https://github.com/flathub/com.github.mtkennerly.ludusavi/blob/master/com.github.mtkennerly.ludusavi.yaml) specifies:

```
  - --filesystem=home
  - --filesystem=/run/media/
  - --filesystem=/media/
  - --filesystem=/var/home/
```

I would have thought that the `home` permission would be enough here, but I found this note:

https://docs.flatpak.org/en/latest/sandbox-permissions.html#filesystem-access

> The `home` permission also has exceptions as it does not grant access to the subdirectories for other applications in `~/.var/app/`.

I've updated the beta to include `--filesystem=~/.var/app/com.heroicgameslauncher.hgl` (based on https://github.com/flatpak/flatpak/issues/2735#issuecomment-469499858). Could you try removing the Flatseal changes and see if the updated beta can find The Witness and The Witcher?

If that works, then I'll add the permissions to the stable Flatpak release as well, and then we can investigate Batman and Into the Breach specifically.

**Comment by djibux** [comment]
Nice! It works for both games. Thanks for the quick fix.

Should I open another issue for Batman and Into the Breach?

**Comment by mtkennerly** [comment]
I think this issue's fine since we're already here :)

If you run these commands and grab the two most recent log files, then I can take a look:

```
flatpak run --env=RUST_LOG=ludusavi=trace com.github.mtkennerly.ludusavi//beta backup --preview "Into the Breach"
```

```
flatpak run --env=RUST_LOG=ludusavi=trace com.github.mtkennerly.ludusavi//beta backup --preview "Batman: Arkham Knight"
```

**Comment by djibux** [comment]
For Into The Breach, it turns out I had not played the game on the Deck yet.  There were no saves… so the behavior is normal!

For Batman, I don’t know what’s happening.  Save files are located in `/run/media/mmcblk0p1/Heroic/BatmanArkhamKnight/BMGame/SaveData` that is scanned by Ludusavi.
[ludusavi_rCURRENT_batman.log](https://github.com/mtkennerly/ludusavi/files/11039630/ludusavi_rCURRENT_batman.log)

**Comment by mtkennerly** [comment]
After `checking`, the next thing  it should log is `found` (file), `looking for files in` (directory), or `excluded`. Since it's not logging any of those, it doesn't think the path exists.

* Does the capitalization match exactly? I see both `BMGame` and `BmGame` on the wiki.
* What if you add `/run/media/mmcblk0p1/Heroic/BatmanArkhamKnight/BMGame/SaveData` in Flatseal? The Flatpak already has `--filesystem=/run/media/`, which should be enough, but it would be good to confirm.

**Comment by djibux** [comment]
Well spotted, the folder was `BmGame`.  I renamed it to `BMGame` and ludusavi finds the saves.  Windows (wine) does not seem to care that the case changed.

**Comment by mtkennerly** [comment]
Yeah, Ludusavi should ignore case for non-native games. It does that for Steam/Proton, but I'll need to add it for Heroic as well. Thanks for helping investigate this :)

**Comment by djibux** [comment]
Well, thank *you*.  You’ve given clear instructions, pinpointed every issue quickly and released fixes.  What else could I ask for!
